# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [84]:
import pandas as pd


# Notebook Presentation

In [85]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [86]:
df = pd.read_csv('googleplaystore.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [87]:
df.sample(5)

App       Category  Rating Reviews  \
358    Firefox Focus: The privacy browser  COMMUNICATION    4.40   36880   
10074                     Virtual Dice EX           GAME    4.30      40   
4520                               Q Call          TOOLS     NaN      20   
5552                             Super AR         FAMILY    3.80     410   
5624        Five Tries At Love Dating Sim         FAMILY    4.20   37165   

       Size    Installs  Type Price Content Rating         Genres  \
358    4.0M  1,000,000+  Free     0       Everyone  Communication   
10074  3.6M      5,000+  Free     0       Everyone          Board   
4520    16M      1,000+  Free     0       Everyone          Tools   
5552    28M     10,000+  Free     0           Teen  Entertainment   
5624    22M  1,000,000+  Free     0           Teen     Simulation   

            Last Updated Current Ver Android Ver  
358         July 6, 2018         5.2  5.0 and up  
10074     March 15, 2016         2.8  3.0 and up  
4520       July 16, 2018       2.100  4.0 and up  
5552   December 18, 2017         2.1  4.0 and up  
5624   February 14, 2018       5.1.0  2.3 and up

In [88]:
df.shape

(10841, 13)

In [89]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [90]:
df.drop(columns=['Last Updated', 'Android Ver'], inplace=True)

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [91]:
df['Rating'].isnull().sum()

1474

In [92]:
df.dropna(inplace=True)
df.shape

(9362, 11)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


In [93]:
df.duplicated().sum()

476

In [94]:
df.drop_duplicates(inplace=True)

If we do this without specifying how to identify duplicates, we see that 3 copies of Instagram are retained because they have a different number of reviews. We need to provide the column names that should be used in the comparison to identify duplicates. For example:

In [95]:
df[df['App'] == 'Instagram']

App Category  Rating   Reviews                Size  \
2545  Instagram   SOCIAL    4.50  66577313  Varies with device   
2604  Instagram   SOCIAL    4.50  66577446  Varies with device   
3909  Instagram   SOCIAL    4.50  66509917  Varies with device   

            Installs  Type Price Content Rating  Genres         Current Ver  
2545  1,000,000,000+  Free     0           Teen  Social  Varies with device  
2604  1,000,000,000+  Free     0           Teen  Social  Varies with device  
3909  1,000,000,000+  Free     0           Teen  Social  Varies with device

In [96]:
df.drop_duplicates(subset=['App', 'Type', 'Price'], inplace=True)

In [97]:
df.shape

(8194, 11)

In [98]:
df['Size'] = df['Size'].replace(r'[KM]+$', '', regex=True)

In [99]:
df['Size'] = pd.to_numeric(df['Size'], errors='coerce')

In [100]:
average_size_by_category = df[df['Size'].notna()].groupby('Category')['Size'].mean()

def impute_size(row):
    if pd.isna(row['Size']):
        return average_size_by_category.get(row['Category'], 0)
    return row['Size']

df['Size'] = df.apply(impute_size, axis=1)

In [101]:
df['Installs'] = df['Installs'].str.replace(',', '').str.replace('+', '').astype(int)

<ipython-input-101-9affa6ed14c9>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['Installs'] = df['Installs'].str.replace(',', '').str.replace('+', '').astype(int)


In [102]:
df.Reviews = df.Reviews.astype('int')

In [103]:
df.Price = df.Price.str.replace('$', '').astype(float)

<ipython-input-103-1a72ac3cc7c3>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df.Price = df.Price.str.replace('$', '').astype(float)


In [104]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8194 entries, 0 to 10840
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8194 non-null   object 
 1   Category        8194 non-null   object 
 2   Rating          8194 non-null   float64
 3   Reviews         8194 non-null   int64  
 4   Size            8194 non-null   float64
 5   Installs        8194 non-null   int64  
 6   Type            8194 non-null   object 
 7   Price           8194 non-null   float64
 8   Content Rating  8194 non-null   object 
 9   Genres          8194 non-null   object 
 10  Current Ver     8194 non-null   object 
dtypes: float64(3), int64(2), object(6)
memory usage: 768.2+ KB


What else should I know about the data?



So we can see that 13 different features were originally scraped from the Google Play Store.



Obviously, the data is just a sample out of all the Android apps. It doesn't include all Android apps of which there are millions.

I’ll assume that the sample is representative of the App Store as a whole. This is not necessarily the case as, during the web scraping process, this sample was served up based on geographical location and user behaviour of the person who scraped it - in our case Lavanya Gupta.

The data was compiled around 2017/2018. The pricing data reflect the price in USD Dollars at the time of scraping. (developers can offer promotions and change their app’s pricing).

I’ve converted the app’s size to a floating-point number in MBs. If data was missing, it has been replaced by the average size for that category.

The installs are not the exact number of installs. If an app has 245,239 installs then Google will simply report an order of magnitude like 100,000+. I’ve removed the '+' and we’ll assume the exact number of installs in that column for simplicity.

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [105]:
df.sort_values('Rating', ascending=False)

App            Category  Rating  \
8327  The Divine Feminine App: the DF App           LIFESTYLE    5.00   
6807                            Jabbla BT               TOOLS    5.00   
9194         UP EB Bill Payment & Details              SOCIAL    5.00   
9218                  EB Cash Collections            BUSINESS    5.00   
7270                        CF Townsville  HEALTH_AND_FITNESS    5.00   
...                                   ...                 ...     ...   
4127                    Speech Therapy: F              FAMILY    1.00   
8820                       DS Creator 2.0               TOOLS    1.00   
8875                       DT future1 cam               TOOLS    1.00   
5151                        Clarksburg AH             MEDICAL    1.00   
7144                        CB Mobile Biz             FINANCE    1.00   

      Reviews  Size  Installs  Type  Price Content Rating            Genres  \
8327        8  6.70      1000  Free   0.00       Everyone         Lifestyle   
6807        3  9.30       100  Free   0.00       Everyone             Tools   
9194        3  2.40        50  Free   0.00           Teen            Social   
9218        1  4.30         5  Free   0.00       Everyone          Business   
7270        4 14.00       100  Free   0.00       Everyone  Health & Fitness   
...       ...   ...       ...   ...    ...            ...               ...   
4127        1 16.00        10  Paid   2.99       Everyone         Education   
8820        2  4.40       500  Free   0.00       Everyone             Tools   
8875        1 24.00        50  Free   0.00       Everyone             Tools   
5151        1 28.00        50  Free   0.00       Everyone           Medical   
7144        3  8.40       500  Free   0.00       Everyone           Finance   

       Current Ver  
8327         0.0.4  
6807           1.0  
9194           2.0  
9218          1.10  
7270         4.2.2  
...            ...  
4127           1.0  
8820  2.0.180226.1  
8875           3.1  
5151   300000.0.81  
7144      4.4.1255  

[8194 rows x 11 columns]

The main drawback of this approach is that, apps with very few good ratings will also come at the top, leading to bias in our analysis. A better approach would be to consider only those app which are installed by more than a certain threshold and then sort by rating.

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [106]:
df.sort_values('Size', ascending=False).head()

App            Category  \
9170                     Stickman Legends: Shadow Wars                GAME   
4690                                        Vi Trainer  HEALTH_AND_FITNESS   
5862                             Miami crime simulator                GAME   
8409  Car Crash III Beam DH Real Damage Simulator 2018                GAME   
1793                 Mini Golf King - Multiplayer Game                GAME   

      Rating  Reviews   Size  Installs  Type  Price Content Rating  \
9170    4.40    38419 100.00   1000000  Paid   0.99   Everyone 10+   
4690    3.60      124 100.00      5000  Free   0.00       Everyone   
5862    4.00   254518 100.00  10000000  Free   0.00     Mature 17+   
8409    3.60      151 100.00     10000  Free   0.00       Everyone   
1793    4.50   531458 100.00   5000000  Free   0.00       Everyone   

                Genres Current Ver  
9170            Action      2.3.24  
4690  Health & Fitness    2.0.0.11  
5862            Action         2.0  
8409            Racing        1.03  
1793            Sports      3.04.1

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [108]:
df.sort_values('Reviews', ascending=False).head()

App       Category  Rating  \
2544                                  Facebook         SOCIAL    4.10   
336                         WhatsApp Messenger  COMMUNICATION    4.40   
2545                                 Instagram         SOCIAL    4.50   
335   Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
1670                            Clash of Clans           GAME    4.60   

       Reviews  Size    Installs  Type  Price Content Rating         Genres  \
2544  78158306 17.53  1000000000  Free   0.00           Teen         Social   
336   69119316 13.64  1000000000  Free   0.00       Everyone  Communication   
2545  66577313 17.53  1000000000  Free   0.00           Teen         Social   
335   56642847 13.64  1000000000  Free   0.00       Everyone  Communication   
1670  44891723 98.00   100000000  Free   0.00   Everyone 10+       Strategy   

             Current Ver  
2544  Varies with device  
336   Varies with device  
2545  Varies with device  
335   Varies with device  
1670           10.322.16

If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app!

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [112]:
ratings = df['Content Rating'].value_counts()
ratings

Everyone           6616
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content Rating, dtype: int64

In [113]:
import plotly.express as px

In [114]:
fig = px.pie(labels=ratings.index,
values=ratings.values,
title="Content Rating",
names=ratings.index,
)
fig.update_traces(textposition='outside', textinfo='percent+label')

fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning:

Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.



To create a donut 🍩 chart, we can simply add a value for the hole argument:

In [115]:
fig = px.pie(labels=ratings.index,
values=ratings.values,
title="Content Rating",
names=ratings.index,
hole=0.6,
)
fig.update_traces(textposition='inside', textfont_size=15, textinfo='percent')

fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning:

Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.



# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [116]:
df[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         697
5000         424
10000        987
50000        456
100000      1095
500000       503
1000000     1414
5000000      607
10000000     937
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [118]:
df.sort_values('Price', ascending=False).head(20)[['App', 'Price']]

App  Price
4367        I'm Rich - Trump Edition 400.00
4197          most expensive app (H) 399.99
5354                  I am Rich Plus 399.99
5359              I am rich(premium) 399.99
5362                   I Am Rich Pro 399.99
4362                      💎 I'm rich 399.99
5358                      I am Rich! 399.99
5356               I Am Rich Premium 399.99
5373              I AM RICH PRO PLUS 399.99
5364  I am rich (Most expensive app) 399.99
5351                       I am rich 399.99
5369                       I am Rich 399.99
5366                       I Am Rich 389.99
5357             I am extremely Rich 379.99
5355                   I am rich VIP 299.99
2253       Vargo Anesthesia Mega App  79.99
2414                    LTC AS Legal  39.99
5360                I am Rich Person  37.99
2301         A Manual of Acupuncture  33.99
5489       AP Art History Flashcards  29.99

What’s going on here? There are 15 I am Rich Apps in the Google Play Store apparently. They all cost \$300.00 or more, which is the main point of the app. The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for $999.90. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well. The high installation numbers are likely gamed by making the app was available for free at some point to get reviews and appear more legitimate.

### The most expensive apps sub $250

In [120]:
df = df[df['Price'] < 250]
df.sort_values('Price', ascending=False).head(20)[['App', 'Price']]

App  Price
2253                           Vargo Anesthesia Mega App  79.99
2414                                        LTC AS Legal  39.99
5360                                    I am Rich Person  37.99
2301                             A Manual of Acupuncture  33.99
2266                                            EMT PASS  29.99
5489                           AP Art History Flashcards  29.99
3039                             Golfshot Plus: Golf GPS  29.99
2298                                  PTA Content Master  29.99
2251    Human Anatomy Atlas 2018: Complete 3D Human Body  24.99
2277   Muscle Premium - Human Anatomy, Kinesiology, B...  24.99
8719                                   DRAGON QUEST VIII  19.99
2297                                Hospitalist Handbook  19.99
7477                          USMLE Step 2 CK Flashcards  19.99
5036                                          NewTek NDI  19.99
7347                             boattheory.ch Full 2018  19.40
5361                              I am Rich Premium Plus  18.99
4203                             The World Ends With You  17.99
6837                                  SkyTest BU/GU Lite  17.99
2256                          2017 EMRA Antibiotic Guide  16.99
10782                            Trine 2: Complete Story  16.99

### Highest Grossing Paid Apps (ballpark estimate)

In [121]:
df['Revenue Estimate'] = df['Price'] * df['Installs']
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN    4.10   
1                                Coloring book moana  ART_AND_DESIGN    3.90   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN    4.70   
3                              Sketch - Draw & Paint  ART_AND_DESIGN    4.50   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN    4.30   

   Reviews  Size  Installs  Type  Price Content Rating  \
0      159 19.00     10000  Free   0.00       Everyone   
1      967 14.00    500000  Free   0.00       Everyone   
2    87510  8.70   5000000  Free   0.00       Everyone   
3   215644 25.00  50000000  Free   0.00           Teen   
4      967  2.80    100000  Free   0.00       Everyone   

                      Genres         Current Ver  Revenue Estimate  
0               Art & Design               1.0.0              0.00  
1  Art & Design;Pretend Play               2.0.0              0.00  
2               Art & Design               1.2.4              0.00  
3               Art & Design  Varies with device              0.00  
4    Art & Design;Creativity                 1.1              0.00

In [124]:
df.sort_values('Revenue Estimate', ascending=False).head(10)[['App', 'Category', 'Price', 'Installs', 'Revenue Estimate']]

App     Category  Price  Installs  \
2241                      Minecraft       FAMILY   6.99  10000000   
4034                  Hitman Sniper         GAME   0.99  10000000   
7417  Grand Theft Auto: San Andreas         GAME   6.99   1000000   
2883            Facetune - For Free  PHOTOGRAPHY   5.99   1000000   
5578        Sleep as Android Unlock    LIFESTYLE   5.99   1000000   
8804            DraStic DS Emulator         GAME   4.99   1000000   
4133                   Weather Live      WEATHER   5.99    500000   
2242     Card Wars - Adventure Time       FAMILY   2.99   1000000   
5631        Five Nights at Freddy's         GAME   2.99   1000000   
8860                    Bloons TD 5       FAMILY   2.99   1000000   

      Revenue Estimate  
2241     69,900,000.00  
4034      9,900,000.00  
7417      6,990,000.00  
2883      5,990,000.00  
5578      5,990,000.00  
8804      4,990,000.00  
4133      2,995,000.00  
2242      2,990,000.00  
5631      2,990,000.00  
8860      2,990,000.00

The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 4 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

If you were to release an app, would you choose to go after a competitive category with many other apps? Or would you target a popular category with a high number of downloads? Or perhaps you can target a category which is both popular but also one where the downloads are spread out among many different apps. That way, even if it’s more difficult to discover among all the other apps, your app has a better chance of getting installed, right? Let’s analyse this with bar charts and scatter plots and figure out which categories are dominating the market.

In [125]:
df.Category.nunique()

33

In [127]:
top10_category = df.Category.value_counts()[:10]
top10_category

FAMILY             1603
GAME                912
TOOLS               718
PRODUCTIVITY        301
PERSONALIZATION     298
FINANCE             296
LIFESTYLE           296
MEDICAL             291
BUSINESS            263
PHOTOGRAPHY         263
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [128]:
bar = px.bar(x = top10_category.index, # index = category name
             y = top10_category.values)

bar.show()

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.



But what if we look at it from a different perspective? What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, we have to group all our apps by category and sum the number of installations:

In [129]:
category_installs = df.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [131]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation='h',
               title='Category Popularity')

h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

Now we see that Games and Tools are actually the most popular categories. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is. Do few apps have most of the downloads or are the downloads spread out over many apps?

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [136]:
cat_app_installs = df.groupby('Category').agg({'App': 'count', 'Installs': 'sum'}).sort_values('Installs', ascending=False)
cat_app_installs

App     Installs
Category                              
GAME                  912  13878762717
COMMUNICATION         256  11038241530
TOOLS                 718   8099224500
PRODUCTIVITY          301   5793070180
SOCIAL                203   5487841475
PHOTOGRAPHY           263   4649143130
FAMILY               1603   4427404490
VIDEO_PLAYERS         148   3926797200
TRAVEL_AND_LOCAL      187   2894859300
NEWS_AND_MAGAZINES    204   2369110650
ENTERTAINMENT         102   2113660000
BOOKS_AND_REFERENCE   169   1665791655
PERSONALIZATION       298   1532352930
SHOPPING              180   1400331540
HEALTH_AND_FITNESS    244   1144006220
SPORTS                260   1096431465
BUSINESS              263    697018120
LIFESTYLE             296    503610120
MAPS_AND_NAVIGATION   118    503267560
FINANCE               296    455249400
WEATHER                72    361096500
EDUCATION             118    352852000
FOOD_AND_DRINK         94    211677750
DATING                134    140912410
ART_AND_DESIGN         60    114228100
HOUSE_AND_HOME         62     97082000
AUTO_AND_VEHICLES      73     53129800
LIBRARIES_AND_DEMO     63     51083000
COMICS                 54     44931100
MEDICAL               291     38252676
PARENTING              50     31116110
BEAUTY                 42     26916200
EVENTS                 45     15949410

In [137]:
scatter = px.scatter(cat_app_installs, # data
                    x='App', # column name
                    y='Installs',
                    title='Category Concentration',
                    size='App',
                    hover_name=cat_app_installs.index,
                    color='Installs')

scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log'))

scatter.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.



# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


In [140]:
df.Genres.value_counts()

Tools                          717
Entertainment                  467
Education                      429
Productivity                   301
Personalization                298
                              ... 
Health & Fitness;Education       1
Music & Audio;Music & Video      1
Arcade;Pretend Play              1
Art & Design;Pretend Play        1
Strategy;Creativity              1
Name: Genres, Length: 114, dtype: int64

If we look at the number of unique values in the Genres column we get 114. But this is not accurate if we have nested data like we do here. We can see this using .value_counts() and looking at the values that just have a single entry. There we see that the semi-colon (;) separates the genre names.

In [149]:
num_genres = df.Genres.str.split(';', expand=True).stack().value_counts()
num_genres[:10]

Tools              718
Education          587
Entertainment      498
Action             304
Productivity       301
Personalization    298
Lifestyle          297
Finance            296
Medical            291
Sports             270
dtype: int64

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [150]:
bar = px.bar(x = num_genres.index[:15], # index = category name
             y = num_genres.values[:15], # count
             title='Top Genres',
             hover_name=num_genres.index[:15],
             color=num_genres.values[:15],
             color_continuous_scale='Agsunset')

bar.update_layout(xaxis_title='Genre',
yaxis_title='Number of Apps',
coloraxis_showscale=False)

bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

Now that we’ve looked at the total number of apps per category and the total number of apps per genre, let’s see what the split is between free and paid apps.

In [151]:
df.Type.value_counts()

Free    7591
Paid     588
Name: Type, dtype: int64

We see that the majority of apps are free on the Google Play Store. But perhaps some categories have more paid apps than others. Let’s investigate. We can group our data first by Category and then by Type. Then we can add up the number of apps per each type. Using as_index=False we push all the data into columns rather than end up with our Categories as the index.

In [154]:
df_free_vs_paid = df.groupby(['Category', 'Type'], as_index=False)['App'].count()
df_free_vs_paid

Category  Type  App
0      ART_AND_DESIGN  Free   57
1      ART_AND_DESIGN  Paid    3
2   AUTO_AND_VEHICLES  Free   72
3   AUTO_AND_VEHICLES  Paid    1
4              BEAUTY  Free   42
..                ...   ...  ...
56   TRAVEL_AND_LOCAL  Paid    8
57      VIDEO_PLAYERS  Free  144
58      VIDEO_PLAYERS  Paid    4
59            WEATHER  Free   65
60            WEATHER  Paid    7

[61 rows x 3 columns]

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [155]:
g_bar = px.bar(df_free_vs_paid,
               x='Category',
               y='App',
               title='Free vs Paid Apps by Category',
               color='Type',
               barmode='group')

g_bar.update_layout(xaxis_title='Category',
                    yaxis_title='Number of Apps',
                    xaxis={'categoryorder':'total descending'},
                    yaxis=dict(type='log'))

g_bar.show()

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.
But this leads to many more questions:

How much should you charge? What are other apps charging in that category?

How much revenue could you make?

And how many downloads are you potentially giving up because your app is paid?

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


In [156]:
box = px.box(df,
             y='Installs',
             x='Type',
             color='Type',
             notched=True,
             points='all',
             title='How Many Downloads are Paid Apps Giving Up?')

box.update_layout(yaxis=dict(type='log'))

box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

But does this mean we should give up on selling a paid app? Let’s see how much revenue we would estimate per category.

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [158]:
df_paid_apps = df[df['Type'] == 'Paid']
box = px.box(df_paid_apps,
             x='Category',
             y='Revenue Estimate',
             title='How Much Can Paid Apps Earn?')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))


box.show()

If an Android app costs $30,000 to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about $20,000. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.